autoscale/src/main/java/com/sequenceiq/periscope/service/YarnInvocationService.java

```diff
@@ -0,0 +1,138 @@
+import com.sequenceiq.periscope.utils.ClusterUtils;
+import com.sequenceiq.periscope.utils.StackResponseUtils;
+import com.sequenceiq.periscope.utils.TimeUtil;
+
+@Service
+public class YarnInvocationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(YarnInvocationService.class);
+
+    @Inject
+    private YarnResponseUtils yarnResponseUtils;
+
+    @Inject
+    private StackResponseUtils stackResponseUtils;
+
+    @Inject
+    private YarnMetricsClient yarnMetricsClient;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private ScalingEventSender eventSender;
+
+    private Cluster cluster;
+
+    private LoadAlert loadAlert;
+
+    private LoadAlertConfiguration loadAlertConfiguration;
+
+    private String policyHostGroup;
+
+    public void pollYarnMetricsAndScaleCluster(Long clusterId, String pollingUserCrn, boolean stopStartEnabled, StackV4Response stackV4Response)
+            throws Exception {
+
+        cluster = clusterService.findById(clusterId);
+        loadAlert = cluster.getLoadAlerts().iterator().next();
+        policyHostGroup = loadAlert.getScalingPolicy().getHostGroup();
+        loadAlertConfiguration = loadAlert.getLoadAlertConfiguration();
+
+        Map<String, String> hostFqdnsToInstanceId = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackV4Response, policyHostGroup);
+        List<String> stoppedHostInstanceIds = stackResponseUtils.getStoppedCloudInstanceIdsInHostGroup(stackV4Response, policyHostGroup);
+
+        int serviceHealthyHostGroupSize = stackResponseUtils.getCloudInstanceIdsWithServicesHealthyForHostGroup(stackV4Response, policyHostGroup).size();
+        int existingHostGroupSize = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackV4Response, policyHostGroup).size();
+
+        int maxResourceValueOffset = loadAlertConfiguration.getMaxResourceValue() - (stopStartEnabled ? serviceHealthyHostGroupSize : existingHostGroupSize);
+        int minResourceValueOffset = serviceHealthyHostGroupSize - loadAlertConfiguration.getMinResourceValue();
+
+        int maxAllowedUpScale = Math.max(maxResourceValueOffset, 0);
+        int maxAllowedDownScale = Math.max(minResourceValueOffset, 0);
+
+        LOGGER.info("Various counts: hostFqdnsToInstanceId={}, servicesHealthyHostFqdnsToInstanceId: {}, stoppedHostFqdnsToInstanceId: {}, " +
+                        "maxResourceValueOffset={}, minResourceValueOffset={}, maxAllowedUpScale={}, maxAllowedDownScale={}",
+                existingHostGroupSize, serviceHealthyHostGroupSize, stoppedHostInstanceIds.size(),
+                maxResourceValueOffset, minResourceValueOffset, maxAllowedUpScale, maxAllowedDownScale);
+
+        Optional<Integer> mandatoryDownScaleCount = Optional.of(maxResourceValueOffset)
+                .filter(mandatoryDownscale -> mandatoryDownscale < 0).map(downscale -> -1 * downscale);
+
+        YarnScalingServiceV1Response yarnResponse =
```

Comment:
mandatoryDownScaleCount represents the maximum number of nodes that we may want to scale down. Can be renamed accordingly. The code above can be simplified for just this.

Reply:
Renamed this